### DISCUSSION TITLE
Types of custom structures and enums

### DISCUSSION THREAD

**Discussion by danmoser** [FreeOpcUa/opcua-asyncio#981] [Q&A] [answered]
Hi folks, 
Is there a way to set the nodeset "HasTypeDefinition" to match the contents of the "Value" field (see the image link)?
Running the `server-custom-structures-and-enums.py` example, all created nodesets have the "BaseDataVariableType"...

For some reason, I'm unable to upload the image here ("Something went really wrong, and we can't process that file"). 
[Figure](http://danmoser.github.io/figs/Types_struct.png)
In example of the figure, I need to have `MyStruct` in the `HasTypeDefinition` field.

I'm running one application that is reading that definition to decode the nodeset contents (if it is a custom one).

Thanks!

**Comment by schroeder-** [comment, marked as answer]
I think you mix something up. `HasTypeDefinition` in the case of a Variable means, you provide a VariableType on the server as Template. In your case you can create a MyStructVariableType and add HasTypeDefinition to your variable.
In the server-custom-structures-and-enums.py like this:
``` Python
    valtype = server.nodes.base_variable_type.add_variable_type(idx, f"{idx}:MyStructVariableType", snode1)
    val = await server.nodes.objects.add_variable(idx, "my_struct", ua.Variant(ua.MyStruct(), ua.VariantType.ExtensionObject))
    val.add_reference(valnode, ua.ObjectIds.HasTypeDefinition, val)
```

**Reply by danmoser** [reply]
Thank you @schroeder- !!
Indeed my question was confusing, but you understood what I needed. I found some typos in your answer, so for reference I'm posting what worked for me:
```
    valtype = await server.nodes.base_variable_type.add_variable_type(idx, f"{idx}:MyStructVariableType", snode1.nodeid)
    val = await server.nodes.objects.add_variable(idx, "my_struct", ua.Variant(ua.MyStruct(), ua.VariantType.ExtensionObject))
    await val.add_reference(valtype, ua.ObjectIds.HasTypeDefinition)
    await val.delete_reference(ua.ObjectIds.BaseDataVariableType, ua.ObjectIds.HasTypeDefinition)
```
